sdk/opendp/whitenoise/synthesizers/dpctgan/dpctgan.py

```diff
@@ -0,0 +1,286 @@
+import numpy as np
+import torch
+from torch import optim
+from torch.nn import functional
+import torch.nn as nn
+import torch.utils.data
+from torch.nn import BatchNorm1d, Dropout, LeakyReLU, Linear, Module, ReLU, Sequential,Sigmoid
+from torch.nn import functional as F
+from opendp.whitenoise.synthesizers.base import SDGYMBaseSynthesizer
+
+import ctgan
+from ctgan.transformer import DataTransformer
+from ctgan.conditional import ConditionalGenerator 
+from ctgan.models import Discriminator, Generator
+from ctgan.sampler import Sampler 
+
+from ctgan import CTGANSynthesizer
+
+import torchdp
+
+
+# custom for calcuate grad_sample for multiple loss.backward()
+def _custom_create_or_extend_grad_sample(
+    param: torch.Tensor, grad_sample: torch.Tensor, batch_dim: int
+) -> None:
+    """
+    Create a 'grad_sample' attribute in the given parameter, or accumulate it
+    if the 'grad_sample' attribute already exists.
+
+    This custom code will not work when using optimizer.virtual_step()
+    """
+    
+    if hasattr(param, "grad_sample"):
+        param.grad_sample = param.grad_sample + grad_sample
```

Comment:
can we look into the approach mentioned where we add the virtual_step attribute for the override to avoid affecting base scenarios?

if hasattr(param, "grad_sample") and getattr(param, "virtual_step", False):
           param.grad_sample = param.grad_sample + grad_sample
  elif if hasattr(param, "grad_sample"):
       param.grad_sample = torch.cat((param.grad_sample, grad_sample), batch_dim)
  else:
        param.grad_sample = grad_sample